core/src/qos_service.rs

```diff
@@ -84,50 +84,57 @@ impl QosService {
         }
     }
 
-    /// Calculate cost of transactions, determine which ones to include in the slot, and
-    /// accumulate costs in the cost tracker.
-    /// Returns a vector of transaction costs, a vector of results indicating which transactions
-    /// were selected, and the number of transactions that were *NOT* selected.
+    /// Calculate cost of transactions, if not already filtered out, determine which ones to
+    /// include in the slot, and accumulate costs in the cost tracker.
+    /// Returns a vector of results containing selected transaction costs, and the number of
+    /// transactions that were *NOT* selected.
     pub fn select_and_accumulate_transaction_costs(
         &self,
         bank: &Bank,
         transactions: &[SanitizedTransaction],
-    ) -> (Vec<TransactionCost>, Vec<transaction::Result<()>>, usize) {
+        pre_results: impl Iterator<Item = transaction::Result<()>>,
+    ) -> (Vec<transaction::Result<TransactionCost>>, usize) {
```

Comment:
Tied together the result and cost returns.

We don't need the `TransactionCost` unless it's a transaction that was selected.